File: solutions/csharp/binary-search/1/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        if (input.Length == 0) return -1;
        int sp = 0;
        int ep = input.Length-1;
        int mp = input.Length / 2;
        while (input[mp] != value)
        {
            if (sp == mp || mp == ep) return -1;
            if (input[sp] < value && input[sp+1]> value) return -1;
            if (value == input[sp]) return sp;
            if (value == input[ep]) return ep;
            if (value < input[mp]) ep = mp;
            else sp = mp;
            mp = (sp + ep) / 2;
        }
        return mp;
    }
}

File: solutions/cpp/binary-search/1/binary_search.cpp
#include "binary_search.h"
using namespace std;
namespace binary_search {
    size_t find(vector<int> data, int item){
        size_t b;
        size_t m;
        size_t e;
        if (data.size()==0)throw domain_error(" ");
        b=0; 
        e=data.size()-1;
        m=(b+e)/2;
        while (true){
            if (item==data[e]) return e;
            if (item==data[m]) return m;
            if (item<data[b]) throw domain_error(" ");
            if (item>data[e]) throw domain_error(" ");
            if (e-b== 1 && item>data[b]&& item<data[e]) throw domain_error(" ");
            if (item<=data[m]) e=m;
            if (item>data[m]) b=m;
            m=(b+e)/2;
        }

    }
}  // namespace binary_search
